File: axelari-main/backend/app/modules/state/session_service.py
from app.core.db import get_db
from sqlalchemy.orm import Session
from app.core.redis_client import get_redis_client
import uuid
import json

class SessionService:
    def __init__(self, db: Session):
        self.db = db
        self.redis = get_redis_client()

    def start_session(self, student_id: str, grade: int, subject: str):
        session_id = str(uuid.uuid4())
        session_data = {
            "student_id": student_id,
            "grade": grade,
            "subject": subject,
            "status": "active",
            "current_question_index": 0
        }
        # Store in Redis (Upstash)
        self.redis.set(f"session:{session_id}", json.dumps(session_data))
        return {"session_id": session_id, "message": "Session started"}

    def get_next_question(self, session_id: str):
        # Retrieve session from Redis
        session_data_str = self.redis.get(f"session:{session_id}")
        if not session_data_str:
            raise ValueError("Session not found")
        
        # Logic to get next question (mock)
        return {
            "session_id": session_id,
            "question": {
                "id": "q1",
                "topic_id": "t1",
                "question_text": "What is 5 + 7?",
                "question_type": "multiple_choice",
                "options": ["10", "11", "12", "13"],
                "difficulty": 1,
                "estimated_time_seconds": 30
            }
        }

    def submit_answer(self, session_id: str, student_id: str, question_id: str, answer_given: str, time_spent: int):
        # Logic to validate answer and update mastery
        # Mock response
        is_correct = (answer_given == "12")
        return {
            "is_correct": is_correct,
            "correct_answer": "12",
            "explanation": "5 + 7 equals 12",
            "updated_mastery": {
                "topic_id": "t1",
                "mastery_score": 85 if is_correct else 70,
                "status": "mastered" if is_correct else "learning",
                "current_difficulty": 2
            },
            "profile_snapshot": {
                "processing_speed": 60,
                "accuracy_consistency": 70,
                "memory_retention": 65
            },
            "intervention": {
                "needs_intervention": False,
                "intervention_type": "none",
                "message": None
            }
        }

File: axelari-main/axelari1-main/src/services/sessionService.ts
import { httpGet, httpPost } from './httpClient';
import { FASTAPI_BASE } from './api';

export interface SessionStartRequest {
    student_id: string;
    grade: number;
    subject: string;
}

export interface SessionStartResponse {
    session_id: string;
    message: string;
}

export interface Question {
    id: string;
    topic_id: string;
    text: string; // Changed from question_text to match backend
    question_type: string;
    options: any[]; // each option should have { id: string; text: string }
    difficulty: number;
    estimated_time_seconds: number;
}

export interface NextQuestionResponse {
    session_id: string;
    question: Question;
}

export interface SubmitAnswerRequest {
    student_id: string;
    question_id: string;
    answer_given: string; // matches option.id
    time_spent_seconds: number;
}

export interface SubmitAnswerResponse {
    is_correct: boolean;
    correct_answer: string | null;
    explanation: string | null;
    updated_mastery: {
        topic_id: string;
        mastery_score: number;
        status: string;
        current_difficulty: number;
    };
    profile_snapshot: {
        processing_speed: number;
        accuracy_consistency: number;
        memory_retention: number;
    };
    intervention: {
        needs_intervention: boolean;
        intervention_type: string;
        message: string | null;
    };
}

export const startSession = async (
    request: SessionStartRequest
): Promise<SessionStartResponse> => {
    return httpPost(`${FASTAPI_BASE}/session/start`, request);
};

export const getNextQuestion = async (
    sessionId: string
): Promise<NextQuestionResponse> => {
    return httpGet(`${FASTAPI_BASE}/session/${sessionId}/next-question`);
};

export const submitAnswer = async (
    sessionId: string,
    request: SubmitAnswerRequest
): Promise<SubmitAnswerResponse> => {
    return httpPost(`${FASTAPI_BASE}/session/${sessionId}/submit-answer`, request);
};
